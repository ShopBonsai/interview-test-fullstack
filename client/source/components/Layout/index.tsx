import * as React from "react";
import ProductsList from "../Products";
import Toolbar from "./Toolbar";

const Layout: React.FC = () => {
  return (
    <>
      <Toolbar />
      <ProductsList />
    </>
  );
};

export default Layout;
