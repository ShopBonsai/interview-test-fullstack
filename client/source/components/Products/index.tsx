import React from "react";
import {
  CardTitle,
  CardSubtitle,
  CardText,
  Button,
  CardBody,
  Media,
} from "reactstrap";
import "./styles.css";
import { Spinner } from "../UI/Spinner";
import { IProduct } from "../../../../interfaces/IProduct";
import { timerEnd, timerStart } from "../../utils/getServerLoad";
import { useProducts } from "../../context/products-context";
import { useCart } from "../../context/shoppingcart-context";

const ProductsList: React.FC = () => {
  const {
    filteredProducts,
    loading,
    error,
    checkStock,
    updateStock,
  } = useProducts();

  const { addToCart } = useCart();

  const onClickHandler = (id: string) => {
    if (checkStock(id)) {
      const product = updateStock(id);
      if (product) {
        addToCart(product);
      }
    }
  };

  if (loading) {
    timerStart();
    return <Spinner />;
  }

  if (error || !filteredProducts) {
    timerEnd();
    return (
      <div>
        <h3>Could not fetch data!</h3>
      </div>
    );
  }

  if (filteredProducts.length < 1) {
    timerEnd();
    return (
      <div>
        <h3>No products available</h3>
      </div>
    );
  }

  timerEnd();
  return (
    <div className="product-list">
      {filteredProducts.map((product: IProduct) => {
        const {
          color,
          description,
          quantity,
          image,
          name,
          price,
          size,
        } = product;
        return quantity > 0 ? (
          <Media key={product.id} className="product-card">
            <Media left href="#">
              <Media object src={image} alt="Product image cap" />
            </Media>
            <CardBody>
              <CardTitle style={{ fontWeight: 600 }}>{name}</CardTitle>
              <CardTitle>Price: {price}</CardTitle>
              <CardSubtitle>Color: {color}</CardSubtitle>
              <CardSubtitle>Size: {size}</CardSubtitle>
              <CardSubtitle>Stock: {quantity}</CardSubtitle>
              <CardText>Details: {description}</CardText>
              <Button
                color="primary"
                size="lg"
                block
                onClick={() => onClickHandler(product.id)}
              >
                Add to Cart
              </Button>
            </CardBody>
          </Media>
        ) : null;
      })}
    </div>
  );
};

export default ProductsList;
