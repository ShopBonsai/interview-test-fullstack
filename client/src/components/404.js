import React from 'react';
import { Link } from "react-router-dom";

const ErrorPage = (props) =>{
  return (
    <div className="error-page">
      <div className="error-page__content">
        <h1>Oops!</h1>
        <p>The page you’re looking for can’t be found. Go back to your previous page or use the search bar to try again.</p>

        <div className="actions">
          <Link to="/store" className="text-primary">Go to Store</Link>
        </div>
      </div>
    </div>
  );
}

export default ErrorPage;