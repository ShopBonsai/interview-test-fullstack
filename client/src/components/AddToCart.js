import React from 'react';
import { Form, Button, Spinner } from 'reactstrap';
import { Mutation } from 'react-apollo';
import { gql } from 'apollo-boost';

const ADD_TO_CART = gql`
	mutation addToCart($id: String!, $quantity: Int!) {
		addToCart(product: { id: $id, quantity: $quantity }) {
			success
			message
			totalNumberOfCartItems
		}
	}
`;

const AddToCart = ({ children, productId, quantity, updateCart }) => {
	const handleUpdateCart = (res) => {
		// TODO: check if not successful, implemement error toast/message
		if (res.data.addToCart.success) {
			updateCart(res.data.addToCart.totalNumberOfCartItems);
		}
	};

	return (
		<Form>
			{children}
			<Mutation
				mutation={ADD_TO_CART}
				variables={{ id: productId, quantity: quantity }}
			>
				{(AddToCart, { loading, error }) => {
					// TODO: implement better error handling
					if (error) console.log(error);

					return (
						<Button
							color='primary'
							size='lg'
							block
							onClick={() => AddToCart().then((res) => handleUpdateCart(res))}
							disabled={loading}
							style={{
								padding: loading ? '.30rem 1rem' : '.5rem 1rem',
							}}
						>
							{loading ? <Spinner color='light' /> : 'Add to cart'}
						</Button>
					);
				}}
			</Mutation>
		</Form>
	);
};

export default AddToCart;
