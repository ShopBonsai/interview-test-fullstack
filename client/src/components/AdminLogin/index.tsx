import * as React from "react";
import { gql, useMutation } from "@apollo/client";

import Form from "./Form";
import { isLoggedInVar } from "../../apollo/variables";
import "./styles.css";

export const LOGIN_ADMIN = gql`
  mutation LoginAdmin($email: String!, $password: String!) {
    loginAdmin(email: $email, password: $password) {
      email
    }
  }
`;

export default function Login() {
  const [login, { error }] = useMutation<any, any>(LOGIN_ADMIN, {
    onCompleted() {
      isLoggedInVar(true);
    },
  });

  if (error) {
    return <p>An error occurred</p>;
  }

  return (
    <div className="admin-login">
      <Form login={login} />
    </div>
  );
}
