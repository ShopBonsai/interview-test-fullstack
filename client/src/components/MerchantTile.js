import React from 'react';
import { Container, Row, Col, CardTitle, CardText, Media } from 'reactstrap';

import { ProductCard } from './ProductCard';

const styles = {
    container: {
        boxShadow: '5px 5px 10px rgb(136 10 142 / 40%)',
        borderRadius: '10px',
        margin: '1rem',
        width: '95%',
    },
    left: {
        padding: '2rem',
    },
};

export const MerchantTile = ({ data }) => {
    const { merchant, companyDescription, logo, publishedState, products } =
        data;

    if (!publishedState) {
        // we only want to render a given merchant if they are published
        return null;
    }

    return (
        <>
            <Container style={styles.container} fluid={true}>
                <Row style={styles.row}>
                    <Col style={styles.left} sm="4">
                        <Row>
                            <Media
                                object
                                src={logo}
                                alt=" Merchant logo"
                                className="icon"
                            />
                            <CardTitle className="text-title">
                                {merchant}
                            </CardTitle>
                        </Row>
                        <CardText>
                            About {merchant}: {companyDescription}
                        </CardText>
                    </Col>
                    <Col sm="7">
                        {products &&
                            products.length > 0 &&
                            products.map((product) => (
                                <ProductCard key={product.id} data={product} />
                            ))}
                    </Col>
                </Row>
            </Container>
        </>
    );
};
