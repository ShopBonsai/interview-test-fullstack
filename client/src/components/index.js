export { Login } from "./Login";
export { NavHeader } from "./Nav";
export { ProductsList } from "./Products";
export { Signup } from "./Signup";
