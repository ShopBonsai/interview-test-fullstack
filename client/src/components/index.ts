export { default as Products } from './Products';
export { default as Nav } from './Nav';
export { default as Cart } from './Cart';
