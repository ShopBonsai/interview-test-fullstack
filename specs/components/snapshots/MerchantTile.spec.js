import React from 'react';
import renderer from 'react-test-renderer';

import { MerchantTile } from '../../../client/src/components/MerchantTile';
import { merchantData } from '../fixtures/merchantTile';

describe('Merchant Tile component snapshot tests', () => {
    it('should render correctly', () => {
        const tree = renderer
            .create(<MerchantTile data={merchantData} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should not render if the merchant has not been published', () => {
        const data = {
            ...merchantData,
            publishedState: false,
        };
        const tree = renderer.create(<MerchantTile data={data} />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
