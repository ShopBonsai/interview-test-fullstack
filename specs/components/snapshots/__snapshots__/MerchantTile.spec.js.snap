// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merchant Tile component snapshot tests should not render if the merchant has not been published 1`] = `null`;

exports[`Merchant Tile component snapshot tests should render correctly 1`] = `
<div
  className="container-fluid"
  style={
    Object {
      "borderRadius": "10px",
      "boxShadow": "5px 5px 10px rgb(136 10 142 / 40%)",
      "margin": "1rem",
      "width": "95%",
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col-sm-4"
      style={
        Object {
          "padding": "2rem",
        }
      }
    >
      <div
        className="row"
      >
        <img
          alt=" Merchant logo"
          className="icon media-object"
          src="www.url.com/logo.png"
        />
        <div
          className="text-title card-title"
        >
          Vandelay Industries
        </div>
      </div>
      <p
        className="card-text"
      >
        About 
        Vandelay Industries
        : 
        A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description 
      </p>
    </div>
    <div
      className="col-sm-7"
    >
      <div
        className="product-card media"
      >
        <a
          className="media-left"
          href="#"
        >
          <img
            alt="Product image cap"
            className="img-tile media-object"
            src="www.fakeurl.com/img.png"
          />
        </a>
        <div
          className="card-body"
        >
          <div
            className="text-title card-title"
          >
            product title
          </div>
          <div
            className="card-title"
          >
            Price: $
            19.00
          </div>
          <div
            className="card-subtitle"
          >
            Color: 
            #424242
          </div>
          <div
            className="card-subtitle"
          >
            Size: 
            sm
          </div>
          <div
            className="card-subtitle"
          >
            Quantity: 
            5
          </div>
          <p
            className="card-text"
          >
            Details: 
            words words words words words words words words words words 
          </p>
          <button
            aria-label={null}
            className="btn btn-primary btn-lg btn-block"
            onClick={[Function]}
          >
            Buy
          </button>
        </div>
      </div>
      <div
        className="product-card media"
      >
        <a
          className="media-left"
          href="#"
        >
          <img
            alt="Product image cap"
            className="img-tile media-object"
            src="www.fakeurl.com/img.png"
          />
        </a>
        <div
          className="card-body"
        >
          <div
            className="text-title card-title"
          >
            product title
          </div>
          <div
            className="card-title"
          >
            Price: $
            19.00
          </div>
          <div
            className="card-subtitle"
          >
            Color: 
            #424242
          </div>
          <div
            className="card-subtitle"
          >
            Size: 
            sm
          </div>
          <div
            className="card-subtitle"
          >
            Quantity: 
            5
          </div>
          <p
            className="card-text"
          >
            Details: 
            words words words words words words words words words words 
          </p>
          <button
            aria-label={null}
            className="btn btn-primary btn-lg btn-block"
            onClick={[Function]}
          >
            Buy
          </button>
        </div>
      </div>
      <div
        className="product-card media"
      >
        <a
          className="media-left"
          href="#"
        >
          <img
            alt="Product image cap"
            className="img-tile media-object"
            src="www.fakeurl.com/img.png"
          />
        </a>
        <div
          className="card-body"
        >
          <div
            className="text-title card-title"
          >
            product title
          </div>
          <div
            className="card-title"
          >
            Price: $
            19.00
          </div>
          <div
            className="card-subtitle"
          >
            Color: 
            #424242
          </div>
          <div
            className="card-subtitle"
          >
            Size: 
            sm
          </div>
          <div
            className="card-subtitle"
          >
            Quantity: 
            5
          </div>
          <p
            className="card-text"
          >
            Details: 
            words words words words words words words words words words 
          </p>
          <button
            aria-label={null}
            className="btn btn-primary btn-lg btn-block"
            onClick={[Function]}
          >
            Buy
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
